### PR TITLE
feat(nfts): taikoons upgradeable v2

### PR DESCRIPTION
**Pull request by bearni95**
In this PR:
- [x] Add upgrade deployment schema for Taikoons
- [x] Implement `TaikoonTokenV2` with the following methods:
    - [x] `initializeV2`, to call when upgrading
    - [x] `updateBaseURI` (onlyOwner)
    - [x] `baseURI` (public, view)
- [x] Upgrade `forge` script
- [x] Update token metadata
- [x] Deploy the upgrade on mainnet with the corrected IPFS hashes

### CODE REVIEW

File: packages/nfts/contracts/taikoon/TaikoonTokenV2.sol [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import { ERC721EnumerableUpgradeable } from
+    "@openzeppelin/contracts-upgradeable/token/ERC721/extensions/ERC721EnumerableUpgradeable.sol";
+
+import { MerkleWhitelist } from "./MerkleWhitelist.sol";
+import { IMinimalBlacklist } from "@taiko/blacklist/IMinimalBlacklist.sol";
+
+/// @title TaikoonToken
+/// @dev The Taikoons V1 ERC-721 token
+/// @custom:security-contact security@taiko.xyz
+contract TaikoonTokenV2 is ERC721EnumerableUpgradeable, MerkleWhitelist {
```
Comment (dantaik) [inline packages/nfts/contracts/taikoon/TaikoonTokenV2.sol:15]: Please modify the V1 contract directly!

File: packages/nfts/contracts/taikoon/TaikoonToken.sol [hunk 2]
```diff
@@ -112,4 +112,25 @@ contract TaikoonToken is ERC721EnumerableUpgradeable, MerkleWhitelist {
             _mint(_to, tokenIds[i]);
         }
     }
+
+    /// @dev V2 code additions
+
+    /// @notice Contract initializer
+    /// @param _rootURI Base URI for the token metadata
+    function initializeV2(string memory _rootURI) external {
```
Comment (dantaik) [inline packages/nfts/contracts/taikoon/TaikoonToken.sol:120]: so anyone can call this function?

Comment (dantaik) [reply packages/nfts/contracts/taikoon/TaikoonToken.sol:120]: maybe you missed the `initializer(2)` modifier?

File: packages/nfts/contracts/taikoon/TaikoonToken.sol [hunk 3]
```diff
@@ -112,4 +112,25 @@ contract TaikoonToken is ERC721EnumerableUpgradeable, MerkleWhitelist {
...
+        _baseURIExtended = _rootURI;
+    }
+
+    /// @notice Update the base URI
+    /// @param _rootURI The new base URI
+    /// @dev Only the owner can update the base URI
+    function updateBaseURI(string memory _rootURI) public onlyOwner {
```
Comment (dantaik) [inline packages/nfts/contracts/taikoon/TaikoonToken.sol:127]: If you have this function, I think you can remove `initializeV2`

**Review by dantaik — CHANGES_REQUESTED** [review]

**Review by dantaik — APPROVED** [review]

**Review by bennettyong — APPROVED** [review]

**Pull request merged**